File: api/Enkl.Api/Controllers/TasksController.cs
using System.Security.Claims;
using Enkl.Api.Auth;
using Enkl.Api.Dtos;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Enkl.Api.Controllers;

[ApiController]
[Authorize(Policy = "ProjectMember")]
[Route("api/projects/{projectId:guid}/tasks")]
public class TasksController : ControllerBase
{
    private readonly TaskService _tasks;
    private readonly SseBroadcaster _broadcaster;
    private readonly FormSubmissionService _formSubmissions;

    // FormSubmissionService is injected here rather than into TaskService itself, which would be a
    // straight constructor cycle (FormSubmissionService already depends on TaskService to raise a
    // task from a Form Workflow "action" node) — a controller depending on both is a clean one-way
    // graph, not a cycle, since TaskService itself gains no new dependency.
    public TasksController(TaskService tasks, SseBroadcaster broadcaster, FormSubmissionService formSubmissions)
    {
        _tasks = tasks;
        _broadcaster = broadcaster;
        _formSubmissions = formSubmissions;
    }

    // ARCHITECTURE-REVIEW.md finding 2.2: additive, targeted alternative to pulling every task
    // through GET /api/projects/{id} (ProjectService.GetProjectDetailAsync's one all-in-one graph
    // fetch) — see TaskService.GetTasksPagedAsync's own doc comment. page defaults to 1, pageSize to
    // 50 (clamped [1, 200] in the service) if the caller omits either.
    [HttpGet]
    public async Task<IActionResult> List(Guid projectId, [FromQuery] int page = 1, [FromQuery] int pageSize = 50)
    {
        var result = await _tasks.GetTasksPagedAsync(projectId, page, pageSize);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPost]
    public async Task<IActionResult> Create(Guid projectId, CreateTaskRequest request)
    {
        var result = await _tasks.CreateAsync(projectId, request);
        if (result is null) return BadRequest(new { message = "Invalid column." });
        await BroadcastAsync(projectId, result, "created");
        return Ok(result);
    }

    [HttpPut("{taskId:guid}")]
    public async Task<IActionResult> Update(Guid projectId, Guid taskId, UpdateTaskRequest request)
    {
        var result = await _tasks.UpdateAsync(projectId, taskId, request, User.FindFirstValue("displayName"));
        if (result is null) return NotFound();
        await BroadcastAsync(projectId, result, "updated");
        // Cheap no-op for the overwhelming majority of task updates (an indexed lookup that finds
        // nothing) — only actually does anything when this Task was raised by a Form Workflow
        // "raiseTaskInPortal" action node AND this update just moved it into a Done column. See
        // FormSubmissionService.ResumeIfLinkedTaskDoneAsync's own doc comment for the full shape.
        await _formSubmissions.ResumeIfLinkedTaskDoneAsync(taskId, request.FormClosingNotes);
        // Same shape, for the "In Review" transition — only does anything the first time this Task's
        // AssigneeId goes non-null while its linked submission is still 'submitted'.
        await _formSubmissions.MarkInReviewIfTaskAssignedAsync(taskId);
        return Ok(result);
    }

    // Cheap, single-indexed-lookup check the frontend fires only at the moment a Task is about to
    // move into a Done column, to decide whether to show the optional "Add closing notes?" prompt —
    // deliberately NOT part of TaskDto/GetProjectDetailAsync's own graph fetch, which every task on
    // every board load would otherwise pay for. 404 (not an empty 200) when unlinked, so the frontend
    // can treat "not linked" and "no such task" identically — there's nothing to prompt for either way.
    [HttpGet("{taskId:guid}/form-link")]
    public async Task<IActionResult> GetFormLink(Guid projectId, Guid taskId)
    {
        var submissionId = await _formSubmissions.GetRaisedFromTaskIdAsync(projectId, taskId);
        return submissionId is null ? NotFound() : Ok(new TaskFormLinkDto(submissionId.Value));
    }

    [HttpDelete("{taskId:guid}")]
    public async Task<IActionResult> Delete(Guid projectId, Guid taskId)
    {
        // Grab the key/title before deleting so the "X was deleted" toast can still name it.
        var deleted = await _tasks.GetTaskSummaryAsync(projectId, taskId);
        if (!await _tasks.DeleteAsync(projectId, taskId)) return NotFound();
        if (deleted is not null) await BroadcastAsync(projectId, deleted.Value.TaskId, deleted.Value.Key, deleted.Value.Title, "deleted");
        return NoContent();
    }

    // Best-effort — a notification failure must never fail the mutation itself, so any exception here
    // (e.g. a momentarily broken connection registry) is swallowed rather than surfaced to the caller.
    private async Task BroadcastAsync(Guid projectId, TaskDto task, string changeType) =>
        await BroadcastAsync(projectId, task.Id, task.Key, task.Title, changeType);

    private async Task BroadcastAsync(Guid projectId, Guid taskId, string taskKey, string title, string changeType)
    {
        try
        {
            var memberUserIds = await _tasks.GetProjectMemberUserIdsAsync(projectId);
            var userId = User.UserId();
            var displayName = User.FindFirstValue("displayName") ?? "Someone";
            var clientSessionId = Request.Headers["X-Client-Session-Id"].FirstOrDefault();

            _broadcaster.BroadcastTaskChanged(
                memberUserIds,
                new TaskChangedEventDto(projectId, taskId, taskKey, title, changeType, userId, displayName),
                clientSessionId);
        }
        catch
        {
            // Notification is best-effort — the mutation already succeeded and was already returned/
            // will be returned to the caller regardless.
        }
    }
}

File: php-api/src/Controllers/TasksController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Db\Database;
use Enkl\Api\Realtime\Broadcaster;
use Enkl\Api\Services\FormSubmissionService;
use Enkl\Api\Services\TaskService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/** Ported from Controllers/TasksController.cs. */
final class TasksController extends BaseController
{
    private function service(): TaskService
    {
        return new TaskService(Database::connection());
    }

    private function formSubmissions(): FormSubmissionService
    {
        return new FormSubmissionService(Database::connection());
    }

    public function create(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->create($args['projectId'], $this->body($request));
        if ($result === null) {
            return $this->json($response, ['message' => 'Invalid column.'], 400);
        }
        $this->broadcast($request, $args['projectId'], $result['id'], $result['key'], $result['title'], 'created');
        return $this->json($response, $result);
    }

    public function update(Request $request, Response $response, array $args): Response
    {
        $body = $this->body($request);
        $result = $this->service()->update(
            $args['projectId'],
            $args['taskId'],
            $body,
            $this->callerDisplayName($request)
        );
        if ($result === null) {
            return $this->notFound($response);
        }
        $this->broadcast($request, $args['projectId'], $result['id'], $result['key'], $result['title'], 'updated');
        // Cheap no-op for the overwhelming majority of task updates (an indexed lookup that finds
        // nothing) — only actually does anything when this Task was raised by a Form Workflow
        // "raiseTaskInPortal" action node AND this update just moved it into a Done column. See
        // FormSubmissionService::resumeIfLinkedTaskDone's own doc comment for the full shape.
        // formClosingNotes is not a Task field — it's a pass-through that only matters when this
        // update happens to be the Done-column move for a linked submission (ignored otherwise).
        $this->notifyLinkedFormResumed($args['taskId'], $body['formClosingNotes'] ?? null);
        // Same shape, for the "In Review" transition — only does anything the first time this Task's
        // AssigneeId goes non-null while its linked submission is still 'submitted'.
        try {
            $this->formSubmissions()->markInReviewIfTaskAssigned($args['taskId']);
        } catch (\Throwable) {
            // best-effort, same as notifyLinkedFormResumed
        }
        return $this->json($response, $result);
    }

    /** GET .../tasks/{taskId}/form-link — cheap single-indexed-lookup check the frontend fires only
     * at the moment a Task is about to move into a Done column, to decide whether to show the
     * optional "Add closing notes?" prompt. Deliberately NOT part of the Task's own DTO/the full
     * project-detail graph fetch, which every task on every board load would otherwise pay for. 404
     * (not an empty 200) when unlinked, so the frontend treats "not linked" and "no such task"
     * identically — nothing to prompt for either way. */
    public function formLink(Request $request, Response $response, array $args): Response
    {
        $submissionId = $this->formSubmissions()->getRaisedFromTaskId($args['projectId'], $args['taskId']);
        if ($submissionId === null) {
            return $this->notFound($response);
        }
        return $this->json($response, ['submissionId' => $submissionId]);
    }

    /** Best-effort — a notification failure must never fail the task update itself. Broadcast
     * ownership stays at this controller level (same convention as Portal/Forms controllers) — the
     * service only decides WHO/what to notify. */
    private function notifyLinkedFormResumed(string $taskId, ?string $closingNotes): void
    {
        try {
            $result = $this->formSubmissions()->resumeIfLinkedTaskDone($taskId, $closingNotes);
            $target = $result['decisionNotify'] ?? null;
            if ($target === null) {
                return;
            }
            (new Broadcaster(Database::connection()))->broadcastFormSubmissionDecided(
                $target['userId'], $result['projectId'], $result['submissionId'], $result['formName'] ?? '',
                $target['decision'], $target['displayName'], null
            );
        } catch (\Throwable) {
            // best-effort, see comment above
        }
    }

    public function delete(Request $request, Response $response, array $args): Response
    {
        $service = $this->service();
        // Grab the key/title before deleting so the "X was deleted" toast can still name it.
        $deleted = $service->getTaskSummary($args['projectId'], $args['taskId']);
        if (!$service->delete($args['projectId'], $args['taskId'])) {
            return $this->notFound($response);
        }
        if ($deleted !== null) {
            $this->broadcast($request, $args['projectId'], $deleted['taskId'], $deleted['key'], $deleted['title'], 'deleted');
        }
        return $this->noContent($response);
    }

    /** Best-effort — a notification failure must never fail the mutation itself. */
    private function broadcast(Request $request, string $projectId, string $taskId, string $taskKey, string $title, string $changeType): void
    {
        try {
            $service = $this->service();
            $memberUserIds = $service->getProjectMemberUserIds($projectId);
            $claims = $request->getAttribute('jwtClaims');
            $userId = (string) ($claims->sub ?? '');
            $displayName = $claims->displayName ?? 'Someone';
            $clientSessionId = $request->getHeaderLine('X-Client-Session-Id') ?: null;

            (new Broadcaster(Database::connection()))->broadcastTaskChanged(
                $memberUserIds, $projectId, $taskId, $taskKey, $title, $changeType, $userId, $displayName, $clientSessionId
            );
        } catch (\Throwable) {
            // Notification is best-effort — the mutation already succeeded and its response is already
            // being returned to the caller regardless.
        }
    }
}
